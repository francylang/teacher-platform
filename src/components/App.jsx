import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Nav from './Nav';
import CardContainer from './CardContainer';
import DiscussionForm from './DiscussionForm';
import FilterForm from './FilterForm';
import userLoginData from '../utils/userLoginData';
import { Redirect } from 'react-router';
import { PROD_URL } from '../url.js';

class App extends Component {
  constructor() {
    super();
    this.state = {
      allDiscussions: [],
      discussions: [],
      comments: [],
      showingDiscussions: true,
      showingForm: false,
      showingStandards: false,
      isAuthenticated: false
    };
    this.fetchDiscussions = this.fetchDiscussions.bind(this);
    this.renderDiscussions = this.renderDiscussions.bind(this);
    this.renderForm = this.renderForm.bind(this);
    this.renderStandards = this.renderStandards.bind(this);
    this.renderFilteredDiscussions = this.renderFilteredDiscussions.bind(this);
    this.handleSearch = this.handleSearch.bind(this);
  }

  componentDidMount() {
    this.fetchDiscussions();
    this.fetchComments();
  }

  fetchDiscussions() {
    fetch(`http://localhost:3000/api/v1/discussions`)
      .then((response) => response.json())
      .then((rawDiscussions) => this.cleanDiscussions(rawDiscussions))
      .then((allDiscussions) => {
        this.setState({ allDiscussions, discussions: allDiscussions });
      })
      .catch((error) => console.error({ error }));
  }

  cleanDiscussions(rawDiscussions) {
    return rawDiscussions.map(discussion => {
      return {
        id: discussion.id,
        title: discussion.title,
        body: discussion.body,
        tagId: discussion.tagId,
        tagTitle: discussion.tagTitle,
      };
    });
  }

  fetchComments() {
    fetch(`http://localhost:3000/api/v1/comments`)
      .then((response) => response.json())
      .then((comments) => this.cleanComments(comments))
      .then((comments) => this.setState({ comments }))
      .catch((error) => console.error({ error }));
  }

  cleanComments(rawComments) {
    return rawComments.map(comment => {
      return {
        id: comment.id,
        comment: comment.body,
        discussionId: comment.discussionId
      };
    });
  }

  renderDiscussions() {
    this.setState({
      showingDiscussions: true,
      showingForm: false,
      showingStandards: false,
      discussions: this.state.allDiscussions,
    });
  }

  renderForm() {
    this.setState({
      showingDiscussions: false,
      showingForm: true,
      showingStandards: false,
    });
  }

  renderStandards() {
    this.setState({
      showingDiscussions: false,
      showingForm: false,
      showingStandards: true,
    });
  }

  renderFilteredDiscussions(discussions) {
    this.setState({ discussions });
  }

  handleSearch() {
    this.fetchDiscussions();
  }

  userAuthentication() {
    if (this.state.email === userLoginData.email &&
        this.state.password === userLoginData.password) {
      this.setState({
        isAuthenticated: true
      });
    }
  }

  renderCardContainer() {
    const { showingDiscussions, discussions, allDiscussions, comments } = this.state;
    if (showingDiscussions && discussions) {
      return (
        <CardContainer
          allDiscussions={allDiscussions}
          discussions={discussions}
          comments={comments}
          rendered={showingDiscussions}/>
      );
    }
  }

  render() {
    const { allDiscussions, discussions, comments, showingDiscussions, showingForm, showingStandards } = this.state;
    console.log(this.props.signedInStatus);

    if (this.props.signedInStatus === false) {
      return <Redirect to='/login'/>;
    }

    const showForm = showingForm
      ? <DiscussionForm rendered={showingForm}/> : null;

    const showStandards = showingStandards
      ? <FilterForm
        renderDiscussions={this.renderDiscussions}
        renderFilteredDiscussions={this.renderFilteredDiscussions}
        rendered={showingStandards}
        discussions={discussions} /> : null;

    return (
      <section className="app">
        <article className="main">
          <Nav
            allDiscussions={allDiscussions}
            discussions={discussions}
            renderDiscussions={this.renderDiscussions}
            renderForm={this.renderForm}
            renderStandards={this.renderStandards}
            renderFilteredDiscussions={this.renderFilteredDiscussions}
            handleSearch={this.handleSearch}
          />
          <section className="bottom-main">
            { this.renderCardContainer() }
            { showForm }
            { showStandards }
          </section>
        </article>
      </section>
    );
  }
}

App.propTypes = {

};

export default App;
