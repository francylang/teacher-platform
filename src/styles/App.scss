.app {
  text-align: center;
}

.main {
  display: flex;
  flex-direction: column;
}

.bottom-main {
  display: flex;
  flex-direction: row;
  justify-content: space-around;
  padding: 2%;
  background-color: #F7FAFC;
}
