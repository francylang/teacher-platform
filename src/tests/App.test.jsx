import React from 'react';
import ReactDOM from 'react-dom';
import App from '../components/App';
import { shallow } from 'enzyme';
import testSetup from './testSetup';

describe('<App />', () => {
  let wrapper;
  let mockFn = jest.fn();

  beforeEach(() => {
    wrapper = shallow(<App />);
  });

  it('should exist', () => {
    expect(wrapper).toBeDefined();
  });

  it('should have a false default state', () => {
    expect(wrapper.state().filteredDiscussions).toEqual([]);
    expect(wrapper.state().showingDiscussions).toEqual(true);
    expect(wrapper.state().showingForm).toEqual(false);
    expect(wrapper.state().showingStandards).toEqual(false);
  });

  it('should render a main section', () => {
    const main = wrapper.find('.main');

    expect(main.length).toEqual(1);
  });

  it('should render a bottom main section', () => {
    const bottomMain = wrapper.find('.bottom-main');

    expect(bottomMain.length).toEqual(1);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
