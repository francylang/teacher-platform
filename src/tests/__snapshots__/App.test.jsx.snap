// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Header />,
        <article
          className="main"
>
          <Nav
                    discussions={Array []}
                    handleSearch={[Function]}
                    renderDiscussions={[Function]}
                    renderForm={[Function]}
                    renderStandards={[Function]}
          />
          <section
                    className="bottom-main"
          >
                    <CardContainer
                              comments={Array []}
                              discussions={Array []}
                              rendered={true}
                    />
          </section>
</article>,
      ],
      "className": "app",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Nav
              discussions={Array []}
              handleSearch={[Function]}
              renderDiscussions={[Function]}
              renderForm={[Function]}
              renderStandards={[Function]}
/>,
            <section
              className="bottom-main"
>
              <CardContainer
                            comments={Array []}
                            discussions={Array []}
                            rendered={true}
              />
</section>,
          ],
          "className": "main",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "discussions": Array [],
              "handleSearch": [Function],
              "renderDiscussions": [Function],
              "renderForm": [Function],
              "renderStandards": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <CardContainer
                  comments={Array []}
                  discussions={Array []}
                  rendered={true}
/>,
                null,
                null,
              ],
              "className": "bottom-main",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "comments": Array [],
                  "discussions": Array [],
                  "rendered": true,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              null,
              null,
            ],
            "type": "section",
          },
        ],
        "type": "article",
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Header />,
          <article
            className="main"
>
            <Nav
                        discussions={Array []}
                        handleSearch={[Function]}
                        renderDiscussions={[Function]}
                        renderForm={[Function]}
                        renderStandards={[Function]}
            />
            <section
                        className="bottom-main"
            >
                        <CardContainer
                                    comments={Array []}
                                    discussions={Array []}
                                    rendered={true}
                        />
            </section>
</article>,
        ],
        "className": "app",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Nav
                discussions={Array []}
                handleSearch={[Function]}
                renderDiscussions={[Function]}
                renderForm={[Function]}
                renderStandards={[Function]}
/>,
              <section
                className="bottom-main"
>
                <CardContainer
                                comments={Array []}
                                discussions={Array []}
                                rendered={true}
                />
</section>,
            ],
            "className": "main",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "discussions": Array [],
                "handleSearch": [Function],
                "renderDiscussions": [Function],
                "renderForm": [Function],
                "renderStandards": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <CardContainer
                    comments={Array []}
                    discussions={Array []}
                    rendered={true}
/>,
                  null,
                  null,
                ],
                "className": "bottom-main",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "comments": Array [],
                    "discussions": Array [],
                    "rendered": true,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                null,
                null,
              ],
              "type": "section",
            },
          ],
          "type": "article",
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
